 ***Political & Sentiment Classification***

---


 *Introduction*
 
> The goal of this notebook is to:
1.   Train a model for sentiment analysis based of the title and text description of Youtube videos.
2.   Train a model for political classification based of title and text description of Youtube videos

*Summary*

> Since we are dealing with a NLP problem, we decided to use the pretrained model BERT, more exactly [PolitBERT ](https://huggingface.co/maurice/PolitBERT) (BERT flavour specialized on political speeches, interviews and press briefings of English-speaking politicians), since we only consider videos from News&Politics category.

> We then fine-tuned the model for our specific task, using various datasets to capture the specificity of what we were looking for.



> ***Sentiment Analysis Model***

> For the Sentiment Analysis Model, we used the 
[Twitter and Reddit Sentimental analysis Dataset](https://www.kaggle.com/datasets/cosmos98/twitter-and-reddit-sentimental-analysis-dataset), in order to capture how people are talking online.



> ***Political Classification***

> For the Political Classification , we used the 
[Democrat Vs. Republican Tweets](https://www.kaggle.com/datasets/kapastor/democratvsrepublicantweets), in order to capture how political ideologies and affiliations are expressed online.







# Imports

In [1]:
!pip install -q transformers
import pandas as pd
import numpy as np
import pickle
import matplotlib.pyplot as plt
import seaborn as sns
import re
import copy
from tqdm.notebook import tqdm
import gc

import torch
import torch.nn as nn
import torch.nn.functional as F
from torch import optim
from torch.utils.data import Dataset, DataLoader

from transformers import (
    AutoTokenizer, 
    AutoModel,
    get_linear_schedule_with_warmup
)

from sklearn.metrics import (
    accuracy_score,
    f1_score
)

from sklearn.model_selection import train_test_split

from transformers import BertTokenizer

     |████████████████████████████████| 5.5 MB 7.5 MB/s 
     |████████████████████████████████| 7.6 MB 54.6 MB/s 
     |████████████████████████████████| 182 kB 54.1 MB/s 


# Process the text

In [ ]:
# Pre-process the text.
# Remove punctuation marks and trailing spaces from text.
def clean_and_parse_text(text):
    if type(text) is not str:
      text = ''

    text = text.split()
    text = [x.strip().lower() for x in text]
    text = [x.replace('\n', ' ').replace('\t', ' ') for x in text]
    text = ' '.join(text)
    text = re.sub('([.,!?()])', r' \1 ', text)
    return text

# Get the text from the dataframe and process it.
def get_texts(df):
    texts = df.apply(lambda x: clean_and_parse_text(x))
    texts = texts.values.tolist()
   
    return texts

In [ ]:
# Define the Data used.
class TransformerDataset(Dataset):
  def __init__(self, df, labels=None, set_type=None):
    super(TransformerDataset, self).__init__()

    self.texts = get_texts(df)
    
    self.set_type = set_type
    if self.set_type != 'test':
      self.labels = labels
    
    self.tokenizer = config.TOKENIZER
    self.max_length = config.MAX_LENGTH

  def __len__(self):
      return len(self.texts)
    
  def __getitem__(self, index):
    tokenized = self.tokenizer.encode_plus(
        self.texts[index], 
        max_length=self.max_length,
        pad_to_max_length=True,
        truncation=True,
        return_attention_mask=True,
        return_token_type_ids=False,
        return_tensors='pt'
    )
    input_ids = tokenized['input_ids'].squeeze()
    attention_mask = tokenized['attention_mask'].squeeze()

    # For training, we also need the labels.
    if self.set_type != 'test':
      return {
          'input_ids': input_ids.long(),
          'attention_mask': attention_mask.long(),
          'labels': torch.Tensor(self.labels[index]),
      }

    return {
        'input_ids': input_ids.long(),
        'attention_mask': attention_mask.long(),
    }

# Model Configurations

In [ ]:
# Used parameters to train the models.
class Config:
  def __init__(self):
    super(Config, self).__init__()

    self.SEED = 42
    self.MODEL_PATH = 'maurice/PolitBERT'
    self.NUMBER_POLITICAL_PARTIES = 2
    self.NUMBER_EMOTIONS = 3

    # data
    self.TOKENIZER = AutoTokenizer.from_pretrained(self.MODEL_PATH)
    self.MAX_LENGTH = 320
    self.BATCH_SIZE = 16
    self.VALIDATION_SPLIT = 0.25

     # model
    self.DEVICE = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
    self.FULL_FINETUNING = True
    self.LR = 3e-5
    self.OPTIMIZER = 'AdamW'
    self.CRITERION = 'BCEWithLogitsLoss'
    self.N_VALIDATE_DUR_TRAIN = 3
    self.N_WARMUP = 0
    self.SAVE_BEST_ONLY = True
    self.EPOCHS = 50

config = Config()

# Define the generic model

In [ ]:
# Define the layers of the model. 
class TransformerModel(nn.Module):
  def __init__(self, transformer_model: AutoModel, classes: int):
    super(TransformerModel, self).__init__()
    self.transformer = transformer_model
    self.output = nn.Linear(768, classes)

  def forward(self, input_ids, attention_mask=None, token_type_ids=None):
    _, o2 = self.transformer(
      input_ids=input_ids,
      attention_mask=attention_mask,
      token_type_ids=token_type_ids,
      return_dict=False
    )
    x = self.output(o2)

    return x

# Generic functions for training the model

In [ ]:
device = config.DEVICE
device

In [ ]:
# Evaluate the current model
def val(model, val_dataloader, criterion):
    
    val_loss = 0
    true, pred = [], []
    
    # Set model.eval() every time during evaluation
    model.eval()
    
    for step, batch in enumerate(val_dataloader):
        # Unpack the batch contents and push them to the device (cuda or cpu).
        b_input_ids = batch['input_ids'].to(device)
        b_attention_mask = batch['attention_mask'].to(device)
        b_labels = batch['labels'].to(device)

        # Using torch.no_grad() during validation/inference is faster 
        # since it does not update gradients.
        with torch.no_grad():
            # Forward pass
            logits = model(input_ids=b_input_ids, attention_mask=b_attention_mask)
            
            # Calculate loss
            loss = criterion(logits, b_labels)
            val_loss += loss.item()
             
            # Since we're using BCEWithLogitsLoss, to get the predictions 
            # sigmoid has to be applied on the logits first
            logits = torch.sigmoid(logits)
            
            logits = np.round(logits.cpu().numpy())
            
            labels = b_labels.cpu().numpy()
            
            # The tensors are detached from the gpu and put back on 
            # the cpu, and then converted to numpy in order to 
            # use sklearn's metrics.
            pred.extend(logits)
            true.extend(labels)

    avg_val_loss = val_loss / len(val_dataloader)
    print('Eval Val loss:', avg_val_loss)
    print('Eval Val accuracy:', accuracy_score(true, pred))
    
    
    val_micro_f1_score = f1_score(true, pred, average='micro')
    print('Eval Val micro f1 score:', val_micro_f1_score)
    return val_micro_f1_score

def train(model, train_dataloader, val_dataloader, criterion, optimizer, scheduler, epoch):
    
    # Validate config.N_VALIDATE_DUR_TRAIN times during the training loop
    nv = config.N_VALIDATE_DUR_TRAIN
    temp = len(train_dataloader) // nv
    
    if temp > 100:
      temp = temp - (temp % 100)
    validate_at_steps = [temp * x for x in range(1, nv + 1)]
    
    train_loss = 0
    for step, batch in enumerate(tqdm(train_dataloader, 
                                      desc='Epoch ' + str(epoch))):
        # Set model.eval() every time during training
        model.train()
        
        # Unpack the batch contents and push them to the device (cuda or cpu).
        b_input_ids = batch['input_ids'].to(device)
        b_attention_mask = batch['attention_mask'].to(device)
        b_labels = batch['labels'].to(device)

        # Clear accumulated gradients
        optimizer.zero_grad()

        # Forward pass
        logits = model(input_ids=b_input_ids, attention_mask=b_attention_mask)
        
        # Calculate loss
        loss = criterion(logits, b_labels)
        train_loss += loss.item()

        # Backward pass
        loss.backward()

        # Update weights
        optimizer.step()
        
        # Update scheduler
        scheduler.step()

        if step in validate_at_steps:
            print(f'-- Step: {step}')
            _ = val(model, val_dataloader, criterion)
    
    avg_train_loss = train_loss / len(train_dataloader)
    print('Training loss:', avg_train_loss)


In [ ]:
def run(model, train_dataloader, val_dataloader, model_name):
    # Setting a seed ensures reproducible results.
    # Seed may affect the performance too.
    torch.manual_seed(config.SEED)

    criterion = nn.BCEWithLogitsLoss()
    
    # Define the parameters to be optmized 
    # and add regularization
    if config.FULL_FINETUNING:
        param_optimizer = list(model.named_parameters())
        no_decay = ["bias", "LayerNorm.bias", "LayerNorm.weight"]
        optimizer_parameters = [
            {
                "params": [
                    p for n, p in param_optimizer if not any(nd in n for nd in no_decay)
                ],
                "weight_decay": 0.001,
            },
            {
                "params": [
                    p for n, p in param_optimizer if any(nd in n for nd in no_decay)
                ],
                "weight_decay": 0.0,
            },
        ]
        optimizer = optim.AdamW(optimizer_parameters, lr=config.LR)
    
    num_training_steps = len(train_dataloader) * config.EPOCHS
    scheduler = get_linear_schedule_with_warmup(
        optimizer,
        num_warmup_steps=0,
        num_training_steps=num_training_steps
    )
    
    max_val_micro_f1_score = float('-inf')
    for epoch in range(config.EPOCHS):
        train(model, train_dataloader, val_dataloader, criterion, optimizer, scheduler, epoch)
        val_micro_f1_score = val(model, val_dataloader, criterion)
        print("Epoch " + str(epoch) + "/" + str(config.EPOCHS) + ": F1 Score " + str(val_micro_f1_score))
        if config.SAVE_BEST_ONLY:
            if val_micro_f1_score > max_val_micro_f1_score:
                best_model = copy.deepcopy(model)
                best_val_micro_f1_score = val_micro_f1_score

                torch.save(best_model.state_dict(), model_name + '.pt')

                print(f'--- Best Model. Val loss: {max_val_micro_f1_score} -> {val_micro_f1_score}')
                max_val_micro_f1_score = val_micro_f1_score

    return best_model, best_val_micro_f1_score


# Sentiment Analysis

## Get the training data

In [ ]:
df = pd.read_csv("Reddit_Data.csv")
X_sent = df['clean_comment']
y_sent = pd.DataFrame(df['category'])

## Pre-process the training data

In [ ]:
# Create a column for each sentiment. 
# Each snetiment will be mapped for an individual node in the final layer.
y_sent['-1'] = int(y_sent['category'] == -1)
y_sent['0'] = int(y_sent['category'] == 0)
y_sent['1'] = int(y_sent['category'] == 1)
y_sent = y_sent.drop('category', axis = 1)

## Split dataset into train and test

In [ ]:
X_train_sent, X_test_sent, y_train_sent, y_test_sent = train_test_split(X_sent, y_sent, test_size=0.2, random_state=42)

## Train the model

In [ ]:
transformer_weights = AutoModel.from_pretrained(
  config.MODEL_PATH
)
sentiment_analysis = TransformerModel(transformer_weights, config.NUMBER_EMOTIONS)

In [ ]:
# Run the model on GPU
sentiment_analysis.to(device);

In [ ]:
train_data_sent = TransformerDataset(X_train_sent, np.vstack(y_train_sent.values).astype(np.float))
val_data_sent = TransformerDataset(X_test_sent, np.vstack(y_test_sent.values).astype(np.float))

train_dataloader_sent = DataLoader(train_data_sent, batch_size=config.BATCH_SIZE)
val_dataloader_sent = DataLoader(val_data_sent, batch_size=config.BATCH_SIZE)

best_model, best_val_micro_f1_score = run(sentiment_analysis, train_dataloader_sent, val_dataloader_sent, 'sentiment_model')

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: DeprecationWarning: `np.float` is a deprecated alias for the builtin `float`. To silence this warning, use `float` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.float64` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  """Entry point for launching an IPython kernel.
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: DeprecationWarning: `np.float` is a deprecated alias for the builtin `float`. To silence this warning, use `float` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.float64` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  


Epoch 0:   0%|          | 0/1863 [00:00<?, ?it/s]

/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


-- Step: 600
Eval Val loss: 0.25957285150757675
Eval Val accuracy: 0.8276510067114093
Eval Val micro f1 score: 0.8487958401751506


/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


-- Step: 1200
Eval Val loss: 0.18081660813536063
Eval Val accuracy: 0.8981208053691275
Eval Val micro f1 score: 0.9037656903765691


/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


-- Step: 1800
Eval Val loss: 0.14386240484057067
Eval Val accuracy: 0.9178523489932886
Eval Val micro f1 score: 0.9239773950484391


/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


Training loss: 0.25981179659389325
Eval Val loss: 0.14748089009887927
Eval Val accuracy: 0.9118120805369128
Eval Val micro f1 score: 0.9203825946382863
Epoch 0/50: F1 Score 0.9203825946382863
--- Best Model. Val loss: -inf -> 0.9203825946382863


Epoch 1:   0%|          | 0/1863 [00:00<?, ?it/s]

/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


-- Step: 600
Eval Val loss: 0.13027490015829166
Eval Val accuracy: 0.9284563758389262
Eval Val micro f1 score: 0.933144628656153


/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


-- Step: 1200
Eval Val loss: 0.1305654987666852
Eval Val accuracy: 0.9256375838926174
Eval Val micro f1 score: 0.9312340196474229


/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


-- Step: 1800
Eval Val loss: 0.11681494832512535
Eval Val accuracy: 0.9414765100671141
Eval Val micro f1 score: 0.9442019740817833


/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


Training loss: 0.10067012763151752
Eval Val loss: 0.13876793341553278
Eval Val accuracy: 0.933020134228188
Eval Val micro f1 score: 0.9359586049324643
Epoch 1/50: F1 Score 0.9359586049324643
--- Best Model. Val loss: 0.9203825946382863 -> 0.9359586049324643


Epoch 2:   0%|          | 0/1863 [00:00<?, ?it/s]

/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


-- Step: 600
Eval Val loss: 0.12654743053652587
Eval Val accuracy: 0.9394630872483222
Eval Val micro f1 score: 0.9429877638832863


/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


-- Step: 1200
Eval Val loss: 0.12908332972975184
Eval Val accuracy: 0.9350335570469799
Eval Val micro f1 score: 0.9380198551113497


/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


-- Step: 1800
Eval Val loss: 0.13774068814816248
Eval Val accuracy: 0.936510067114094
Eval Val micro f1 score: 0.93906858139847


/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


Training loss: 0.06113789031985433
Eval Val loss: 0.13144704160833123
Eval Val accuracy: 0.9395973154362416
Eval Val micro f1 score: 0.9425811644754494
Epoch 2/50: F1 Score 0.9425811644754494
--- Best Model. Val loss: 0.9359586049324643 -> 0.9425811644754494


Epoch 3:   0%|          | 0/1863 [00:00<?, ?it/s]

/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


-- Step: 600
Eval Val loss: 0.14087343183266152
Eval Val accuracy: 0.9404026845637584
Eval Val micro f1 score: 0.9434595756110664


/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


-- Step: 1200
Eval Val loss: 0.14586633916713435
Eval Val accuracy: 0.9363758389261745
Eval Val micro f1 score: 0.9394285714285715


/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


-- Step: 1800
Eval Val loss: 0.14429452188428923
Eval Val accuracy: 0.9350335570469799
Eval Val micro f1 score: 0.9394750937332619


/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


Training loss: 0.04270804004866922
Eval Val loss: 0.14233154535009784
Eval Val accuracy: 0.9428187919463087
Eval Val micro f1 score: 0.9455325999463375
Epoch 3/50: F1 Score 0.9455325999463375
--- Best Model. Val loss: 0.9425811644754494 -> 0.9455325999463375


Epoch 4:   0%|          | 0/1863 [00:00<?, ?it/s]

/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


-- Step: 600
Eval Val loss: 0.1480240217034551
Eval Val accuracy: 0.9373154362416107
Eval Val micro f1 score: 0.9407332660612178


/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


-- Step: 1200
Eval Val loss: 0.14401367281814897
Eval Val accuracy: 0.9393288590604026
Eval Val micro f1 score: 0.9431123648330983


/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


-- Step: 1800
Eval Val loss: 0.17118132765436858
Eval Val accuracy: 0.9302013422818792
Eval Val micro f1 score: 0.9331274338659863


/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


Training loss: 0.033255181854729
Eval Val loss: 0.12615013666537914
Eval Val accuracy: 0.9408053691275168
Eval Val micro f1 score: 0.945755192579149
Epoch 4/50: F1 Score 0.945755192579149
--- Best Model. Val loss: 0.9455325999463375 -> 0.945755192579149


Epoch 5:   0%|          | 0/1863 [00:00<?, ?it/s]

/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


-- Step: 600
Eval Val loss: 0.15993051683977885
Eval Val accuracy: 0.9306040268456376
Eval Val micro f1 score: 0.9340231120666488


/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


-- Step: 1200
Eval Val loss: 0.13013128434680601
Eval Val accuracy: 0.941744966442953
Eval Val micro f1 score: 0.9452642876308023


/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


-- Step: 1800
Eval Val loss: 0.15846892555974737
Eval Val accuracy: 0.9332885906040268
Eval Val micro f1 score: 0.9374454075119264


/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


Training loss: 0.025612791469056956
Eval Val loss: 0.16185120900914518
Eval Val accuracy: 0.9361073825503355
Eval Val micro f1 score: 0.9394000402657541
Epoch 5/50: F1 Score 0.9394000402657541


Epoch 6:   0%|          | 0/1863 [00:00<?, ?it/s]

/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


-- Step: 600
Eval Val loss: 0.19050000638120443
Eval Val accuracy: 0.934496644295302
Eval Val micro f1 score: 0.9364844903988183


/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


-- Step: 1200
Eval Val loss: 0.1642497205304116
Eval Val accuracy: 0.9375838926174497
Eval Val micro f1 score: 0.9403786759769034


/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


-- Step: 1800
Eval Val loss: 0.1570112535345534
Eval Val accuracy: 0.9385234899328859
Eval Val micro f1 score: 0.9421020778696791


/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


Training loss: 0.021349093339662655
Eval Val loss: 0.1866218883180971
Eval Val accuracy: 0.9280536912751678
Eval Val micro f1 score: 0.9311039484286866
Epoch 6/50: F1 Score 0.9311039484286866


Epoch 7:   0%|          | 0/1863 [00:00<?, ?it/s]

/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


-- Step: 600
Eval Val loss: 0.21442883589083644
Eval Val accuracy: 0.9343624161073826
Eval Val micro f1 score: 0.9357588776263678


/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


-- Step: 1200
Eval Val loss: 0.16983743382862904
Eval Val accuracy: 0.9385234899328859
Eval Val micro f1 score: 0.9410500872834698


/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


-- Step: 1800
Eval Val loss: 0.18036202749271393
Eval Val accuracy: 0.9339597315436242
Eval Val micro f1 score: 0.9359903381642514


/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


Training loss: 0.017991524802422016
Eval Val loss: 0.14725358019779783
Eval Val accuracy: 0.9375838926174497
Eval Val micro f1 score: 0.9409787577305727
Epoch 7/50: F1 Score 0.9409787577305727


Epoch 8:   0%|          | 0/1863 [00:00<?, ?it/s]

/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


-- Step: 600
Eval Val loss: 0.16450013944433498
Eval Val accuracy: 0.9391946308724832
Eval Val micro f1 score: 0.9412948674941294


/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


-- Step: 1200
Eval Val loss: 0.18379264907242407
Eval Val accuracy: 0.9363758389261745
Eval Val micro f1 score: 0.9395342594456748


/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


-- Step: 1800
Eval Val loss: 0.14738812655509778
Eval Val accuracy: 0.9441610738255034
Eval Val micro f1 score: 0.9457374740089877


/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


Training loss: 0.012949184823148373
Eval Val loss: 0.151411576942134
Eval Val accuracy: 0.9465771812080537
Eval Val micro f1 score: 0.9481590771913353
Epoch 8/50: F1 Score 0.9481590771913353
--- Best Model. Val loss: 0.945755192579149 -> 0.9481590771913353


Epoch 9:   0%|          | 0/1863 [00:00<?, ?it/s]

/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


# Political Analysis

## Get the training data

In [ ]:
df = pd.read_csv('./ExtractedTweets.csv')
X_pol = df['Tweet']
y_pol = pd.DataFrame(df['Party'])

## Pre-process the training data

In [ ]:
# Create a column for each sentiment. 
# Each party will be mapped for an individual node in the final layer.
y_pol['Democrat'] = int(y_pol['Party'] == 'Democrat')
y_pol['Republican'] = int(y_pol['Party'] == 'Republican')
y_pol = y_pol.drop('Party', axis = 1)

## Split dataset into train and test

In [ ]:
X_train_pol, X_test_pol, y_train_pol, y_test_pol = train_test_split(X_pol, y_pol, test_size=0.2, random_state=42)

## Train the model

In [ ]:
transformer_weights = AutoModel.from_pretrained(
  config.MODEL_PATH
)
political_model = TransformerModel(transformer_weights, config.NUMBER_POLITICAL_PARTIES)

In [ ]:
# Run the model on GPU
political_model.to(device);

In [ ]:
train_data_pol = TransformerDataset(X_train_pol, np.vstack(y_train_pol.values).astype(np.float))
val_data_pol = TransformerDataset(X_test_pol, np.vstack(y_test_pol.values).astype(np.float))

train_dataloader_pol = DataLoader(train_data_pol, batch_size=config.BATCH_SIZE)
val_dataloader_pol = DataLoader(val_data_pol, batch_size=config.BATCH_SIZE)

best_model, best_val_micro_f1_score = run(political_model, train_dataloader_pol, val_dataloader_pol, 'political_model')

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: DeprecationWarning: `np.float` is a deprecated alias for the builtin `float`. To silence this warning, use `float` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.float64` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  """Entry point for launching an IPython kernel.
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: DeprecationWarning: `np.float` is a deprecated alias for the builtin `float`. To silence this warning, use `float` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.float64` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  


Epoch 0:   0%|          | 0/4323 [00:00<?, ?it/s]

/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


-- Step: 1400
Eval Val loss: 0.4768966235130832
Eval Val accuracy: 0.7527180198935924
Eval Val micro f1 score: 0.7562626554816315


/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


-- Step: 2800
Eval Val loss: 0.4347767555029937
Eval Val accuracy: 0.7878787878787878
Eval Val micro f1 score: 0.78968517822555


/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


-- Step: 4200
Eval Val loss: 0.40093467222607654
Eval Val accuracy: 0.8074253990284525
Eval Val micro f1 score: 0.8086041401642188


/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


Training loss: 0.4795091863710407
Eval Val loss: 0.40278075822903864
Eval Val accuracy: 0.8079458709229702
Eval Val micro f1 score: 0.809215470890906
Epoch 0/50: F1 Score 0.809215470890906
--- Best Model. Val loss: -inf -> 0.809215470890906


Epoch 1:   0%|          | 0/4323 [00:00<?, ?it/s]

/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


-- Step: 1400
Eval Val loss: 0.4239306390512938
Eval Val accuracy: 0.8056904927133934
Eval Val micro f1 score: 0.80661481974038


/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


-- Step: 2800
Eval Val loss: 0.43602336941081876
Eval Val accuracy: 0.8088711542910016
Eval Val micro f1 score: 0.8092869947377551


/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


-- Step: 4200
Eval Val loss: 0.41410983115724126
Eval Val accuracy: 0.8138445523941708
Eval Val micro f1 score: 0.814379916705229


/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


Training loss: 0.31205980414311396
Eval Val loss: 0.42216208715002146
Eval Val accuracy: 0.8185866296553319
Eval Val micro f1 score: 0.8189705031810296
Epoch 1/50: F1 Score 0.8189705031810296
--- Best Model. Val loss: 0.809215470890906 -> 0.8189705031810296


Epoch 2:   0%|          | 0/4323 [00:00<?, ?it/s]

/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


-- Step: 1400
Eval Val loss: 0.49539794254045205
Eval Val accuracy: 0.813324080499653
Eval Val micro f1 score: 0.8135230493377292


/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


-- Step: 2800
Eval Val loss: 0.5189736401860384
Eval Val accuracy: 0.8141915336571826
Eval Val micro f1 score: 0.8144013880855986


/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


-- Step: 4200
Eval Val loss: 0.5120922061679307
Eval Val accuracy: 0.820205875549387
Eval Val micro f1 score: 0.8206447882029781


/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


Training loss: 0.1995547988671545
Eval Val loss: 0.5203604416018843
Eval Val accuracy: 0.8140180430256766
Eval Val micro f1 score: 0.8143691791519638
Epoch 2/50: F1 Score 0.8143691791519638


Epoch 3:   0%|          | 0/4323 [00:00<?, ?it/s]

/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


-- Step: 1400
Eval Val loss: 0.6085045073133145
Eval Val accuracy: 0.8177770067083044
Eval Val micro f1 score: 0.8179163172657086


/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


-- Step: 2800
Eval Val loss: 0.5556214685947268
Eval Val accuracy: 0.8126879481841314
Eval Val micro f1 score: 0.8130349294471432


/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


-- Step: 4200
Eval Val loss: 0.5490245447570766
Eval Val accuracy: 0.8191071015498497
Eval Val micro f1 score: 0.8192227619708535


/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


Training loss: 0.13566115780088514
Eval Val loss: 0.5328339909114612
Eval Val accuracy: 0.8170830441822808
Eval Val micro f1 score: 0.8172329044383403
Epoch 3/50: F1 Score 0.8172329044383403


Epoch 4:   0%|          | 0/4323 [00:00<?, ?it/s]

/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


-- Step: 1400
Eval Val loss: 0.6636190922852436
Eval Val accuracy: 0.8206685172334027
Eval Val micro f1 score: 0.8208079114015556


/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


-- Step: 2800
Eval Val loss: 0.6492259981038323
Eval Val accuracy: 0.816736062919269
Eval Val micro f1 score: 0.8168729038973054


/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


-- Step: 4200
Eval Val loss: 0.6993811901957727
Eval Val accuracy: 0.813439740920657
Eval Val micro f1 score: 0.8136151999768646


/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


Training loss: 0.09404740448985223
Eval Val loss: 0.6203218648739153
Eval Val accuracy: 0.814943326393708
Eval Val micro f1 score: 0.8150140250426535
Epoch 4/50: F1 Score 0.8150140250426535


Epoch 5:   0%|          | 0/4323 [00:00<?, ?it/s]

/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


-- Step: 1400
Eval Val loss: 0.708330737668323
Eval Val accuracy: 0.8170252139717789
Eval Val micro f1 score: 0.8170488390249545


/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


-- Step: 2800
Eval Val loss: 0.6623085879653547
Eval Val accuracy: 0.8157529493407356
Eval Val micro f1 score: 0.8159157943438783


/usr/local/lib/python3.7/dist-packages/transformers/tokenization_utils_base.py:2310: FutureWarning: The `pad_to_max_length` argument is deprecated and will be removed in a future version, use `padding=True` or `padding='longest'` to pad to the longest sequence in the batch, or use `padding='max_length'` to pad to a max length. In this case, you can give a specific length with `max_length` (e.g. `max_length=45`) or leave max_length to None to pad to the maximal input size of the model (e.g. 512 for Bert).
  FutureWarning,


**Comment**: For both models, the test accuracy and F1-score are quite high (sentiment analysis has an accuracy of 0.94 and for political classification ~0.82)